store/migration/manager.go

```diff
@@ -1,37 +1,87 @@
 package migration
 
 import (
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+
 	"golang.org/x/sync/errgroup"
 
 	"cosmossdk.io/log"
 	"cosmossdk.io/store/v2"
+	"cosmossdk.io/store/v2/commitment"
 	"cosmossdk.io/store/v2/snapshots"
+	"cosmossdk.io/store/v2/storage"
 )
 
 const (
 	// defaultChannelBufferSize is the default buffer size for the migration stream.
 	defaultChannelBufferSize = 1024
 	// defaultStorageBufferSize is the default buffer size for the storage snapshotter.
 	defaultStorageBufferSize = 1024
+
+	migrateChangesetKeyFmt = "m/cs_%x" // m/<version>
 )
 
+// VersionedChangeset is a pair of version and Changeset.
+type VersionedChangeset struct {
+	Version   uint64
+	Changeset *store.Changeset
+}
+
 // Manager manages the migration of the whole state from store/v1 to store/v2.
 type Manager struct {
 	logger           log.Logger
 	snapshotsManager *snapshots.Manager
 
-	storageSnapshotter snapshots.StorageSnapshotter
-	commitSnapshotter  snapshots.CommitSnapshotter
+	stateStorage    *storage.StorageStore
+	stateCommitment *commitment.CommitStore
+
+	db              store.RawDB
+	migratedVersion uint64
+	mtx             sync.Mutex
+	chChangeset     <-chan *VersionedChangeset
+	chDone          chan struct{}
 }
 
 // NewManager returns a new Manager.
-func NewManager(sm *snapshots.Manager, ss snapshots.StorageSnapshotter, cs snapshots.CommitSnapshotter, logger log.Logger) *Manager {
+func NewManager(db store.RawDB, sm *snapshots.Manager, ss *storage.StorageStore, sc *commitment.CommitStore, logger log.Logger) *Manager {
 	return &Manager{
-		logger:             logger,
-		snapshotsManager:   sm,
-		storageSnapshotter: ss,
-		commitSnapshotter:  cs,
+		logger:           logger,
+		snapshotsManager: sm,
+		stateStorage:     ss,
+		stateCommitment:  sc,
+		db:               db,
+	}
+}
+
+// Start starts the whole migration process.
+func (m *Manager) Start(version uint64, chChangeset <-chan *VersionedChangeset, chDone chan struct{}) error {
```

Comment:
Can we add a detailed godoc for `Start` that indicates what's happening. This is a very important and slightly complex method.